crates/kubelet/src/server.rs

```diff
@@ -1,189 +1,113 @@
-async fn get_container_logs<T: Provider + Sync>(
-    provider: &T,
-    req: &Request<Body>,
+async fn get_container_logs<T: 'static + Provider + Send + Sync>(
+    provider: Arc<T>,
     namespace: String,
     pod: String,
     container: String,
-    tail: Option<usize>,
-    follow: bool,
-) -> Response<Body> {
+    opts: LogOptions,
+) -> Result<Response<Body>, Infallible> {
     debug!(
-        "Got container log request for container {} in pod {} in namespace {}. tail: {:?}, follow: {}",
-        container, pod, namespace, tail, follow
+        "Got container log request for container {} in pod {} in namespace {}. Options: {:?}.",
+        container, pod, namespace, opts
     );
-    if namespace.is_empty() || pod.is_empty() || container.is_empty() {
-        return Response::builder()
-            .status(StatusCode::NOT_FOUND)
-            .body(Body::from(format!(
-                "Resource {} not found",
-                req.uri().path()
-            )))
-            .unwrap();
-    }
-    let (sender, log_body) = hyper::Body::channel();
-    let log_sender = LogSender::new(sender, tail, follow);
+    let (sender, log_body) = Body::channel();
+    let log_sender = LogSender::new(sender, opts);
 
     match provider.logs(namespace, pod, container, log_sender).await {
-        Ok(()) => Response::new(log_body),
+        Ok(()) => Ok(Response::new(log_body)),
         Err(e) => {
             error!("Error fetching logs: {}", e);
-            let mut res = Response::new(Body::from(format!("Server error: {}", e)));
-            *res.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
             if e.is::<NotImplementedError>() {
-                res = Response::new(Body::from("Not Implemented"));
-                *res.status_mut() = StatusCode::NOT_IMPLEMENTED;
+                return_with_code(
+                    StatusCode::NOT_IMPLEMENTED,
+                    format!("Logs not implemented in provider."),
+                )
+            } else {
+                return_with_code(
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    format!("Server error: {}", e),
+                )
             }
-            res
         }
     }
 }
+
 /// Run a pod exec command and get the output
 ///
 /// Implements the kubelet path /exec/{namespace}/{pod}/{container}
-fn post_exec<T: Provider>(_provider: &T, _req: &Request<Body>) -> Response<Body> {
-    let mut res = Response::new(Body::from("Not Implemented"));
-    *res.status_mut() = StatusCode::NOT_IMPLEMENTED;
-    res
+async fn post_exec<T: 'static + Provider + Send + Sync>(
+    _provider: Arc<T>,
+    _namespace: String,
+    _pod: String,
+    _container: String,
+) -> Result<Response<Body>, Infallible> {
+    return_with_code(
+        StatusCode::NOT_IMPLEMENTED,
+        format!("Exec not implemented."),
+    )
+}
+
+fn return_with_code(code: StatusCode, body: String) -> Result<Response<Body>, Infallible> {
```

Comment:
Why return a result? So as to not have to wrap it in an `Ok` at the call site?

Reply:
Yes, I wanted to match the return type of the calling functions. This is the type Warp is expecting back from the handler.